// parse/lib/react-native/LiveQuerySubscription.js
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _CoreManager = _interopRequireDefault(require("./CoreManager"));
var _promiseUtils = require("./promiseUtils");
var Subscription = function () {
  function Subscription(id, query, sessionToken) {
    (0, _classCallCheck2.default)(this, Subscription);
    this.id = id;
    this.query = query;
    this.sessionToken = sessionToken;
    this.subscribePromise = (0, _promiseUtils.resolvingPromise)();
    this.unsubscribePromise = (0, _promiseUtils.resolvingPromise)();
    this.subscribed = false;
    var EventEmitter = _CoreManager.default.getEventEmitter();
    this.emitter = new EventEmitter();
    this.on = this.emitter.on;
    this.emit = this.emitter.emit;
    this.on('error', function () {});
  }
  (0, _createClass2.default)(Subscription, [{
    key: "unsubscribe",
    value: function () {
      var _this = this;
      return _CoreManager.default.getLiveQueryController().getDefaultLiveQueryClient().then(function (liveQueryClient) {
        _this.emit('close');
        return liveQueryClient.unsubscribe(_this);
      });
    }
  }]);
  return Subscription;
}();
var _default = Subscription;
exports.default = _default;
